Makefile

```diff
@@ -47,6 +47,17 @@ unit-test: go-generate ## Run unit tests
 			-count=1 \
 			./...
 
+
+.PHONY: functional-test
+functional-test: 
```

Comment:
Not sure the best way to approach this. Ideally you could just run `make functional-test` and it would 1) build the docker image against the latest changes and 2) run the functional tests (by passing the `-tags=functional` flag).

If we naively do

```
.PHONY: build-image
build-image: go-generate
	docker build -t openfga/openfga:latest .

.PHONY: functional-test
functional-test: build-image
	go test -v ...
```
then every time we invoke it we will have to go through the docker build process, which doesn't cache the `go build ...` step. This is pretty time costly, and in a CI/CD environment we pre-build the image anyways.

I'm wondering if it's ok to just assume the image is pre-built and if you want to run functional tests locally you'll have to make sure you do so. What do you think 🤷 ?

Reply:
Both of these choices are not ideal. Do you think it is not enough to run it against `main`?

Reply:
So I originally added the `build-image` Makefile target and had it as a pre-req for this target, but I really didn't like that because every time you run `make functional-test` it does a `docker build ...` which has all of the docker build output, and it takes longer than it probably needs.

The problem is that the `build-image` target shouldn't probably be run if there aren't any sources that have changed that would affect the built image. So to improve that experience we'd want to only have the `build-image` target run if any of the Go source files have changed or any of the artifacts included in the container.

We could do that, it just requires a good chunk of more work to orchestrate that through the Makefile (but it is what make was made for, haha).

WDYT, should we bite the bullet and do ^^?

Reply:
> The problem is that the `build-image` target shouldn't probably be run if there aren't any sources that have changed that would affect the built image.

For local development, I will always have to build the image, no? Because presumably I changed something and I want to see if the tests still pass.

Reply:
@miparnisari 

> For local development, I will always have to build the image, no? Because presumably I changed something and I want to see if the tests still pass.

Most of the time, yes. But not always. If none of the Go files or artifacts that go inside the container have changed, then you don't need to rebuild the image to run functional tests.

> Can you remind me why we did it like this? I forgot :/

Because this gives us the highest fidelity functional test coverage. This approach allows us to test the entrypoint(s) of the container as well as the server code embedded in it. If we only tested the server implementation without spinning up the container, then we could easily foobar the Dockerfile and introduce something unintended. This approach allows us to test the whole integration end to end with very little overhead (you just have to build the container).